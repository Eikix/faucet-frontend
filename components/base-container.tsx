import { FC, PropsWithChildren } from "react";
import { cn } from "@/lib/utils";

interface BaseContainerProps {
  className?: string;
}

export const BaseContainer: FC<PropsWithChildren<BaseContainerProps>> = ({ children, className }) => {
  return (
    <div
      className={cn("flex flex-col justify-center bg-white w-full py-11 md:h-96 rounded-md my-12 sm:my-16", className)}
      style={{
        backgroundImage: `url("/assets/background.svg")`,
        backgroundSize: "cover",
        backgroundPosition: "center",
      }}
    >
      {children}
    </div>
  );
};
