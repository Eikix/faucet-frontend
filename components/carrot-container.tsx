import { FC, PropsWithChildren } from "react";

export const CarrotContainer: FC<PropsWithChildren> = ({ children }) => (
  <div className="flex flex-col justify-center items-center my-16">
    <div className="w-full sm:w-fit text-center flex flex-col justify-center items-center">{children}</div>
  </div>
);
