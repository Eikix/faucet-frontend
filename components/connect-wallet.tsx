import { ConnectButton, ConnectEmbed, useConnectModal } from "thirdweb/react";
import { KAKAROT_SEPOLIA, client, recommendedWallets, wallets } from "@/lib/thirdweb-client";
import { WALLET_MODAL_OPTIONS } from "@/lib/constants";
import { useFaucet } from "@/hooks/useFaucet";
import { Button } from "@/components/ui/button";
import { Avatar, AvatarFallback, AvatarImage } from "@/components/ui/avatar";

export const ConnectWallet = () => {
  const { wallet } = useFaucet();
  const { size, title, titleIcon, showThirdwebBranding } = WALLET_MODAL_OPTIONS;
  const prettyWallet = wallet?.address.slice(0, 6) + "..." + wallet?.address.slice(-4);

  return (
    <div>
      <ConnectButton
        client={client}
        chain={KAKAROT_SEPOLIA}
        wallets={wallets}
        recommendedWallets={recommendedWallets}
        showAllWallets={false}
        theme={WALLET_MODAL_OPTIONS.theme}
        connectModal={{
          size,
          title,
          titleIcon,
          showThirdwebBranding,
        }}
        connectButton={{
          style: {
            display: "flex",
            width: "350px",
            padding: "12px 10px",
            justifyContent: "center",
            alignItems: "center",
            gap: "6px",
            borderRadius: "6px",
            background: "linear-gradient(180deg, rgba(255, 255, 255, 0.12) 0%, rgba(255, 255, 255, 0) 100%), #FF7600",
            boxShadow: "0px 1px 2px 0px rgba(255, 149, 73, 0.31), 0px 0px 0px 1px #FF7600",
            color: "#fff",
            fontSize: "14px",
          },
        }}
        appMetadata={{
          name: "Kakarot Faucet",
          url: "https://sepolia-faucet.kakarot.org/",
          description: "The fast, native faucet to kickstart your journey in the Kakarot ecosystem.",
          logoUrl: "https://sepolia-faucet.kakarot.org/assets/kakarot-logo.svg",
        }}
        detailsButton={{
          render: () => (
            <Button className="w-full space-x-6 items-center text-[#878794] mt-6" variant="outline">
              <div className="flex space-x-1">
                <Avatar>
                  <AvatarImage src={`https://effigy.im/a/${wallet?.address}.png`} />
                  <AvatarFallback>{wallet?.address}</AvatarFallback>
                </Avatar>
                <span>{prettyWallet}</span>
              </div>
              <span className="text-[#FF7600] text-xs">Change Wallet</span>
            </Button>
          ),
        }}
      />
    </div>
  );
};
