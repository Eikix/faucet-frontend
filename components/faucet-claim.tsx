import Image from "next/image";
import { useActiveWallet } from "thirdweb/react";
import { InfoIcon } from "lucide-react";
import { cn } from "@/lib/utils";
import { KAKAROT_SEPOLIA } from "@/lib/thirdweb-client";
import { FaucetStatsResponse } from "@/lib/types";
import { Button } from "@/components/ui/button";

import metamaskLogo from "@/public/assets/metamask.png";
import cooldownCarrot from "@/public/assets/cooldown-carrot.svg";

interface FaucetClaimProps {
  isProcessing: boolean;
  isCooldown: boolean;
  available: string;
  handleClaim: () => void;
  faucetStats?: FaucetStatsResponse;
}

export const FaucetClaim = ({ isCooldown, isProcessing, available, handleClaim, faucetStats }: FaucetClaimProps) => {
  const wallet = useActiveWallet();
  const activeChain = wallet?.getChain();
  const isMetaMask = wallet?.id === "io.metamask";

  const convertSecondsToTime = (seconds: number) => {
    const hours = String(Math.floor(seconds / 3600)).padStart(2, "0");
    const minutes = String(Math.floor((seconds % 3600) / 60)).padStart(2, "0");
    const remainingSeconds = String(seconds % 60).padStart(2, "0");

    if (hours !== "00") return `${hours}:${minutes}:${remainingSeconds} hours`;
    if (minutes !== "00") return `${minutes}:${remainingSeconds} minutes`;
    return `${remainingSeconds} seconds`;
  };

  return (
    <div className="flex flex-col justify-center items-center my-16 sm:my-24">
      <div className="w-full sm:w-fit text-center flex flex-col justify-center items-center">
        {!isProcessing && isCooldown ? (
          <Image src={cooldownCarrot} alt="Cooldown Carrot" />
        ) : (
          <h2 className="text-5xl md:text-7xl leading-tight text-[#878794]">{available}</h2>
        )}
        <Button
          onClick={handleClaim}
          disabled={isProcessing}
          variant={!isProcessing && isCooldown ? "cooldown" : "main"}
          className={cn("mt-6 w-full", !isProcessing && isCooldown && "hidden")}
        >
          {isProcessing ? "Claiming..." : isCooldown ? "Cooldown" : "Claim"}
        </Button>
        {isMetaMask && activeChain?.id !== KAKAROT_SEPOLIA.id && (
          <Button
            variant="outline"
            className="mt-6 w-full space-x-3"
            onClick={() => wallet.switchChain(KAKAROT_SEPOLIA)}
          >
            <Image src={metamaskLogo} alt="metamask" width={16} height={16} />
            <span className="text-[#ff4500]">Add to Metamask</span>
          </Button>
        )}
        {!isProcessing && isCooldown && (
          <div className="flex flex-row items-center justify-center my-4">
            <p className="leading-5 [&:not(:first-child)]:mt-4 text-[#878794] max-w-[350px]">
              You&apos;re on a cooldown period! Try the Kakarot faucet again in{" "}
              {convertSecondsToTime(faucetStats?.timeLeftInS ?? 0)}.
            </p>
          </div>
        )}
      </div>
    </div>
  );
};
