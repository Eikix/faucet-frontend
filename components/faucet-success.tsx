import Image from "next/image";
import Link from "next/link";

import successCarrot from "@/public/assets/success-carrot.svg";
import { Button } from "@/components/ui/button";
import { KKRT_EXPLORER } from "@/lib/constants";

interface FaucetSuccessProps {
  navigateToClaim: () => void;
  txHash: string;
}

export const FaucetSuccess = ({ navigateToClaim, txHash }: FaucetSuccessProps) => (
  <div className="flex flex-col justify-center items-center my-16 sm:my-24">
    <div className="w-full sm:w-fit text-center flex flex-col justify-center items-center">
      <Image src={successCarrot} alt="Success Carrot" />
      <Button onClick={navigateToClaim} variant="main" className="mt-6 w-full">
        Back to Claim
      </Button>
      <Link href={`${KKRT_EXPLORER}/tx/${txHash}`} rel="noopener noreferrer" target="_blank" className="w-full">
        <Button className="space-x-2 w-full mt-2" variant="outline" size="withIcon">
          <span>View on Explorer</span>
          <Image src="/assets/link-icon.svg" alt="Docs" width={16} height={16} />
        </Button>
      </Link>

      <div className="flex flex-row items-center justify-center my-4">
        <p className="leading-5 [&:not(:first-child)]:mt-4 text-[#878794] max-w-[300px]">
          You just got some testnet ETH! Your wallet should reflect this transaction soon!
        </p>
      </div>
    </div>
  </div>
);
