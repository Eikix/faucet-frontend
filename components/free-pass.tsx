import Image from "next/image";
import { Button } from "@/components/ui/button";
import {
  Dialog,
  DialogClose,
  DialogContent,
  DialogDescription,
  DialogFooter,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
} from "@/components/ui/dialog";

import dummyFreePass from "@/public/assets/dummy-free-pass.svg";
import mintingIcon from "@/public/assets/mining-icon.svg";
import linkIcon from "@/public/assets/link-icon.svg";
import xIcon from "@/public/assets/x-icon-inverted.svg";

export const FreePass = () => {
  return (
    <div
      className="flex flex-col justify-center items-center bg-white w-full py-16 px-3 rounded-md"
      style={{
        backgroundImage: `url("/assets/background.svg")`,
        backgroundSize: "cover",
        backgroundPosition: "right",
      }}
    >
      <div className="flex flex-col justify-center items-center text-center max-w-xl">
        <h1 className="scroll-m-20 text-3xl md:text-4xl font-medium tracking-tight lg:text-[52px]">
          Claim your Free Pass
        </h1>
        <p className="leading-7 [&:not(:first-child)]:mt-6  text-[#878794]">
          Welcome to Kakraot beta testnet phase to commemorate this event. <br />
          Claim your free OG Pass by Kakarot Labs.
        </p>
        <Image src={dummyFreePass} width={400} height={400} alt="Free Pass" className="mt-12" />
        <Button variant="main" className="mt-4 md:mt-8 w-full max-w-[400px]">
          Claim your Pass
        </Button>
        <Button variant="outline" className="mt-4 w-full max-w-[400px] text-[#878794]">
          <Image src={mintingIcon} alt="minting" width={24} height={24} priority className="w-[30px] h-6" />
          <span>Minting in Progress</span>
        </Button>
        <Button variant="success" className="mt-4 w-full max-w-[400px]">
          Claimed
        </Button>
        <ClaimedModal />
      </div>
    </div>
  );
};

const ClaimedModal = () => (
  <Dialog>
    <div className="flex w-full space-x-3  max-w-[400px]">
      <Button variant="success" className="mt-4 w-full">
        Claimed
      </Button>
      <DialogTrigger asChild>
        <Button variant="outline" className="mt-4 w-full text-[#878794] gap-1">
          <span>Tx Hash</span>
          <Image src={linkIcon} alt="minting" width={20} height={20} priority />
        </Button>
      </DialogTrigger>
    </div>
    <DialogContent className="sm:max-w-lg">
      <DialogHeader className="flex flex-col items-center">
        <DialogTitle className="text-[#ff5400] text-2xl sm:text-3xl leading-9 font-medium">
          Congratulations!
        </DialogTitle>
        <DialogDescription className="text-center">
          Welcome to Kakraot beta testnet phase to commemorate this event. <br />
          Claim your free OG Pass by Kakarot Labs.
        </DialogDescription>
      </DialogHeader>
      <Image src={dummyFreePass} width={400} height={200} alt="Free Pass" className="w-full" />
      <DialogFooter className="sm:justify-start">
        <Button variant="outline" className="mt-4 w-full gap-1 !bg-black !text-white">
          <span>Share on</span>
          <Image src={xIcon} alt="minting" width={20} height={20} priority />
        </Button>
        <DialogClose asChild>
          <Button variant="outline" className="mt-4 w-full text-[#ff4500]">
            Go to faucet
          </Button>
        </DialogClose>
      </DialogFooter>
    </DialogContent>
  </Dialog>
);
