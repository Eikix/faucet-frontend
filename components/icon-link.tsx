import Image from "next/image";
import Link from "next/link";

interface IconLinkProps {
  src: string;
  href: string;
  name: string;
  className?: string;
}

export const IconLink = ({ src, href, name, className = "" }: IconLinkProps) => {
  return (
    <Link className={className} rel="noopener noreferrer" target="_blank" href={href}>
      <Image src={src} alt={`${name} Logo`} width={24} height={24} priority className="w-[30px] h-6" />
    </Link>
  );
};
