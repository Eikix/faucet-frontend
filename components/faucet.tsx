import Image from "next/image";
import Link from "next/link";
import Confetti from "react-confetti";
import { useEffect, useState } from "react";
import { useActiveAccount, useBlockNumber, useWalletBalance } from "thirdweb/react";
import { toast } from "sonner";

import { KAKAROT_SEPOLIA, client } from "@/lib/thirdweb-client";
import { CONFETTI_COLORS } from "@/lib/constants";
import { useWindowSize } from "@/hooks/useWindowSize";
import { useFaucetJob } from "@/queries/useFaucetJob";
import { useFaucetStats } from "@/queries/useFaucetStats";
import { useFaucetBalance } from "@/queries/useFaucetBalance";
import { useClaimFunds } from "@/mutations/useClaimFunds";
import { FaucetClaim } from "@/components/faucet-claim";
import { FaucetSuccess } from "@/components/faucet-success";
import { TextPair } from "@/components/text-pair";
import { Button } from "@/components/ui/button";

export const Faucet = () => {
  const wallet = useActiveAccount();

  const { mutate: claimFunds, isPending, data: claimJobID } = useClaimFunds();
  const { data: faucetJob, isError } = useFaucetJob(claimJobID?.jobID ?? "");
  const { data: faucetStats } = useFaucetStats(wallet?.address as string);
  const { data: faucetBalance, refetch: refetchFaucet } = useFaucetBalance();
  const { width: windowWidth } = useWindowSize();
  const { refetch: refetchWallet } = useWalletBalance({
    chain: KAKAROT_SEPOLIA,
    address: wallet?.address as string,
    client,
  });

  const [isProcessing, setIsProcessing] = useState(isPending);
  const [isClaimed, setIsClaimed] = useState(false);

  const blockNumber = useBlockNumber({ client, chain: KAKAROT_SEPOLIA });
  const available = `${faucetStats?.dripAmountInEth ?? 0.001} ETH`;

  const isCooldown = faucetStats?.timeLeftInS !== 0 || faucetStats?.canClaim === false;

  const handleClaim = () => {
    setIsProcessing(true);
    claimFunds({ walletAddress: wallet?.address as string });
  };

  useEffect(() => {
    if (faucetJob && faucetJob[0].status === "completed") {
      toast.success("Claimed successfully!");
      setIsProcessing(false);
      setIsClaimed(true);
      refetchFaucet();
      refetchWallet();
    }
  }, [faucetJob, refetchFaucet, refetchWallet]);

  useEffect(() => {
    if (isError) {
      toast.error("Failed to claim funds. Please try again later.");
      setIsProcessing(false);
    }
  }, [isError]);

  return (
    <main className="flex flex-col items-center">
      <div
        className="flex flex-col bg-white w-full py-6 px-3 sm:px-10 lg:px-20 rounded-md mb-12"
        style={{
          backgroundImage: `url("/assets/background.svg")`,
          backgroundSize: "cover",
          backgroundPosition: "right",
        }}
      >
        <Confetti
          colors={CONFETTI_COLORS}
          run={isClaimed}
          numberOfPieces={500}
          recycle={false}
          width={windowWidth}
          hidden={!isClaimed}
        />

        <div className="flex flex-col sm:flex-row gap-4 sm:gap-0 justify-between">
          <DetailAndText title="Faucet Balance" text={`${faucetBalance?.faucetBalanceInEth.substring(0, 6) ?? 0}ETH`} />
          <DetailAndText title="Block Number" text={blockNumber?.toString() ?? "0x"} />
        </div>

        {isClaimed ? (
          <FaucetSuccess navigateToClaim={() => setIsClaimed(false)} />
        ) : (
          <FaucetClaim
            isCooldown={isCooldown}
            isProcessing={isProcessing}
            available={available}
            handleClaim={handleClaim}
            faucetStats={faucetStats}
          />
        )}
      </div>
      <TextPair
        heading="Need more testnet ETH?"
        description="Reach out to us on Discord and raise a ticket if you need large amount of testnet ETH."
      />
      <Link href="https://discord.gg/kakarotzkevm" rel="noopener noreferrer" target="_blank">
        <Button className="space-x-2 max-w-[120px] mt-6" variant="outline" size="withIcon">
          <span>Reach Out</span>
          <Image src="/assets/link-icon.svg" alt="Docs" width={16} height={16} />
        </Button>
      </Link>
    </main>
  );
};

const DetailAndText = ({ title, text }: { title: string; text: string }) => (
  <h4 className="text-[#878794] faucetDetails px-3 py-2 rounded-sm">
    {title}: <span className="text-[#f54400]">{text}</span>
  </h4>
);
