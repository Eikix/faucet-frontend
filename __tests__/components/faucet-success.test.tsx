import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import { FaucetSuccess } from "@/components/faucet-success";

jest.mock("@/lib/constants", () => ({
  KKRT_EXPLORER: "https://explorer.kakarot.io",
}));

describe("FaucetSuccess", () => {
  const mockNavigateToClaim = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders without crashing", () => {
    render(<FaucetSuccess navigateToClaim={mockNavigateToClaim} txHash="0x12345" />);
    expect(screen.getByText("Cha Ching!")).toBeInTheDocument();
  });

  it("calls navigateToClaim function when 'Back to Claim' button is clicked", () => {
    render(<FaucetSuccess navigateToClaim={mockNavigateToClaim} txHash="0x12345" />);
    fireEvent.click(screen.getByText("Back to Claim"));
    expect(mockNavigateToClaim).toHaveBeenCalled();
  });

  it("displays correct link to explorer with transaction hash", () => {
    render(<FaucetSuccess navigateToClaim={mockNavigateToClaim} txHash="0x12345" />);
    const explorerLink = screen.getByText("View on Explorer").closest("a");
    expect(explorerLink).toHaveAttribute("href", "https://explorer.kakarot.io/tx/0x12345");
  });

  it("displays success message correctly", () => {
    render(<FaucetSuccess navigateToClaim={mockNavigateToClaim} txHash="0x12345" />);
    expect(
      screen.getByText("You just got some testnet ETH! Your wallet should reflect this transaction soon!")
    ).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<FaucetSuccess navigateToClaim={mockNavigateToClaim} txHash="0x12345" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
