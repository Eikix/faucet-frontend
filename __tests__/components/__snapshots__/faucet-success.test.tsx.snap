// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FaucetSuccess matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col justify-center items-center my-16 sm:my-24"
  >
    <div
      class="w-full sm:w-fit text-center flex flex-col justify-center items-center"
    >
      <h2
        class="text-3xl md:text-5xl text-[#878794] mb-2"
      >
        Cha Ching!
      </h2>
      <img
        alt="Success Carrot"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 primaryButton text-white font-normal h-10 px-4 py-3 mt-6 w-full"
      >
        Back to Claim
      </button>
      <a
        class="w-full"
        href="https://explorer.kakarot.io/tx/0x12345"
        rel="noopener noreferrer"
        target="_blank"
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground font-normal p-3 rounded-md space-x-2 w-full mt-2"
        >
          <span>
            View on Explorer
          </span>
          <img
            alt="Docs"
            data-nimg="1"
            decoding="async"
            height="16"
            loading="lazy"
            src="/assets/link-icon.svg"
            style="color: transparent;"
            width="16"
          />
        </button>
      </a>
      <div
        class="flex flex-row items-center justify-center my-4"
      >
        <p
          class="leading-5 [&:not(:first-child)]:mt-4 text-[#878794] max-w-[300px]"
        >
          You just got some testnet ETH! Your wallet should reflect this transaction soon!
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
